src/nvim/ex_cmds.c

```diff
@@ -3333,9 +3333,16 @@ static buf_T *do_sub(exarg_T *eap, proftime_T timeout)
 
   // Check for a match on each line.
   linenr_T line2 = eap->line2;
+  // Condition: If preview, then go on if one of the following holds:
+  // * EITHER we don't yet have enough matches for the preview window
+  //    matched_lines.size < (size_t)p_cwh
+  // * OR we're not yet done looking through the lines shown in the
+  //   curren window
+  //     lnum <= curwin->w_botline
```

Comment:
Could that part of the condition be stored in a local variable?

Reply:
Yes I'd think so. I'll try :)

Reply:
Wait, sorry, I did not yet have my 2 mugs of coffee. That condition contains `lnum` which is the running variable of the loop, it changes on every iteration.

We _could_ save `curwin->w_botline - lnum` in a variable `x`, and then compare `lnum - eap->line1` to `x`. Would that be better? Note that lnum is sometimes changed inside the loop itself, so we could not just increment another variable instead of comparing to `lnum - eap->line1`. Otoh, we can save `eap->line1` in a local variable.

So this would become something like

```C
  linenr_T line1 = eap->line1;
  linenr_T to_bot = curwin->w_botline - line1;
  for (linenr_T lnum = line1;
       lnum <= line2 && !got_quit && !aborting()
       && (!preview || matched_lines.size < (size_t)p_cwh
           || lnum - line1 <= to_bot);
       lnum++) 
```

where we'd need something to account for possible negativity of `to_bot`. Is that worthwile, shall I pursue?

Reply:
no, it's ok. Thanks :)

Reply:
We could put an `if (...) break;` at the top of the loop body to cut down the condition size. Sounds like pretty bad practice to me though.